Execution structure (spin-off of #427)

mxgrey opened the pull request:
This PR is spun out of #427 and focuses entirely on changes to the execution structure (Context -> Execution -> Node relationship).

This is the most crucial change for the async execution to be feasible. Up until now, rclrs has had a loose coupling between the executor and nodes. That means nodes could be added to or removed from the executor freely. This comes with the following disadvantages:
* It's possible to add the same node to multiple executors and spin them both at the same time. We have protections in place so that this won't violate any contract with rcl, but it would be a confusing arrangement.
* We need many layers of mutexes, arcs, and weaks to manage the loose ownership.
* Users don't have a clear order in which to invoke these three top-level objects.
* For async execution, nodes will need a way of issuing jobs to their executor at any time, which isn't possible if they can be orphaned or bounce around between executors.

Due to the last bullet point in particular, I'm proposing a rigid procedure in which:
* A user creates a Context
* The Context can create one or more Executors
* Each Executor can create one or more Nodes

The [minimum publisher example](https://github.com/mxgrey/ros2_rust/blob/0874d8d4bf822075d076294c14b0b4e46416b7d2/examples/minimal_pub_sub/src/minimal_publisher.rs#L4-L7) shows a good demonstration of this new procedure.

In this structure, Nodes are no longer allowed to be removed from an Executor or move between Executors. That's a very unlikely use case to begin with, so I don't think there's a downside to this restriction. After #421 we'll be able to introduce custom flavors of Executors, so if anyone really cares about moving their node around between executors, it would be possible to introduce a custom hierarchical executor that can move nodes between other executors.

mxgrey commented on rclrs/src/context.rs:
```diff
@@ -250,14 +273,14 @@ mod tests {
     fn test_create_context() -> Result<(), RclrsError> {
         // If the context fails to be created, this will cause a panic
-        let _ = Context::new(vec![])?;
+        let _ = Context::new(vec![], InitOptions::default())?;
```
I think for building the `Context` there's enough of a difference between the args and the init options that I want to treat them differently.

In the vast majority of cases, users will want to simply pass along the CLI args from the environment, so they'll want to use [`from_env(InitOptions)`](https://github.com/ros2-rust/ros2_rust/pull/428/files#diff-d3f692f0c6fe18e3daae7e4325ea66cbafb4250b7546ccbeab0a7868e77f4484R161-R163) or [`default_from_env()`](https://github.com/ros2-rust/ros2_rust/pull/428/files#diff-d3f692f0c6fe18e3daae7e4325ea66cbafb4250b7546ccbeab0a7868e77f4484R167-R169). Trying to bundle that all up into a single InitOptions struct doesn't strike me as beneficial.

And right now the only field in `InitOptions` is `domain_id: Option<usize>` which doesn't really afford us a nice way to implicitly cast it into an `InitOptions`. For example, if a user just passes in `Some(10)` or `None` and we choose to implicitly cast that into `InitOptions`, that could be implemented just fine for now.. but it would become ambiguous and confusing if we ever add another `Option<T>` field to `InitOptions`.

mxgrey commented on rclrs/src/executor.rs:
```diff
@@ -1,48 +1,58 @@
 /// Single-threaded executor implementation.
-pub struct SingleThreadedExecutor {
-    nodes_mtx: Mutex<Vec<Weak<Node>>>,
+pub struct Executor {
```
> I wonder if by renaming a SingleThreadedExecutor to just Executor it might be a bit confusing if in the future we want to add a new type of executor.

In #421 this `Executor` class will become an opaque wrapper for a user-selected executor. That means the user will be able to choose, at runtime, any kind of executor they want (single-threaded, multi-threaded, custom implementation, whatever) and the rest of the rclrs code will still just see it as the same `Executor`.

`Context::create_basic_executor` is used in #421 to create the single-threaded async executor that we provide out-of-the-box. By introducing `Context::create_basic_executor` it in this PR, users will not experience any API breakage when #421 is merged.

mxgrey commented on rclrs/src/parameter/service.rs:
```diff
@@ -375,12 +378,12 @@ mod tests {
-            .namespace(ns)
-            .build()
+        let client = executor
+            .create_node(NodeOptions::new("client").namespace(ns))
```
I'm open to this. I wrote `NodeOptions` long before I wrote the new `_Options` structs for the primitive types, so this one is a bit out of date already.

mxgrey replied:
Done in [6059506](https://github.com/ros2-rust/ros2_rust/pull/428/commits/605950695c62eed54f02b03ea9f4f6a21f0e28a5). Now `&str` can be implicitly cast into a `NodeOptions`, just like `PrimitiveOptions` in #429.

mxgrey commented on rclrs/src/parameter/service.rs:
```diff
@@ -564,14 +567,14 @@ mod tests {
             .create_client::<SetParametersAtomically>("/get_set/node/set_parameters_atomically")?;
 
         try_until_timeout(|| {
+            println!(" >> testing services");
```
Whoops, removed by [259fcb3](https://github.com/ros2-rust/ros2_rust/pull/428/commits/259fcb374e05624a6cb6061bbe1d10f87bb1f7ed)

esteve commented:
@mxgrey could you revert the version bump? I prefer doing it before a new release separate from other PRs

mxgrey commented:
@esteve I can certainly do that but I should caution that without the version bump cargo will not recognize that it needs to update its lockfile. I ran into on issue on my machine where compiling the examples was failing because cargo insisted on using an outdated version of rclrs that it had cached. I had to delete my colcon build and install directories and delete the target directories in the cargo packages before it would compile.

Bumping the version number across all the `Cargo.toml`s as I did in the last commit should spare others from needing to manually purge their build artifacts. Do you still want me to revert it?

mxgrey commented:
FWIW another way to avoid the problem that I just described is for us to put a workspace `Cargo.toml` in the root of the git repo. That's something I've been wanting to do for a long time now, but I've been holding off on it in case it's controversial for any reason. `colcon-cargo` should be able to support it by now, so that shouldn't be a blocker at least.

esteve commented on examples/minimal_client_service/src/minimal_client_async.rs:
```diff
@@ -1,12 +1,10 @@
 #[tokio::main]
 async fn main() -> Result<(), Error> {
-    let context = rclrs::Context::new(env::args())?;
+    let mut executor = rclrs::Context::default_from_env()?.create_basic_executor();
```
@mxgrey how would a custom executor work here? It'd take a `Context` as a argument, right? I'd remove the `create_basic_executor` function and be slightly more verbose (`SingleThreadedExecutor::new(&context)`) to show how the API works and basically to not make any distinction between the executors.

mxgrey replied:
As discussed in the Working Group meeting, I've change this PR to use an extension trait to provide the `Context::create_basic_executor` method: https://github.com/ros2-rust/ros2_rust/pull/428/commits/a75cd24455d0722a0c3b98c16cdba40f15a90e11

At the same time I've simplified all the tests and demos to just do `use rclrs::*;` since that's probably how most users will want to use rclrs anyway.

mxgrey commented:
@esteve After giving it more thought I've reverted the version bump. It would fix the problem for people pulling this PR, but then we'd need to keep bumping the version number with each of the subsequent PRs that are lined up right now or else the problem will keep happening.

I think gathering all the crates up into one cargo workspace is the right solution here. How would you feel if I open up a separate PR for that?

mxgrey commented:
I should bring attention to a last minute change that I just introduced: [b465f6f](https://github.com/ros2-rust/ros2_rust/pull/428/commits/b465f6f74eba71270e843d690e44f81ac78ca873)

While developing the workers PR #446 I realized that a multi-threaded executor could accumulate multiple `RclrsError`s across its different threads within one spin. I don't want to discard any of those errors just for the sake of returning a single `Result<(), RclrsError>` from `Executor::spin`, so in #446 I changed `Executor::spin` to return a `Vec<RclrsError>` which will be empty if no errors exist.

To allow this to funnel into a single `Result<(), RclrsError>` instance when users want it to, I introduced the trait method [`RclrsErrorFilter::first_error`](https://github.com/mxgrey/ros2_rust/blob/b465f6f74eba71270e843d690e44f81ac78ca873/rclrs/src/error.rs#L385-L386) which can be applied to `Vec<RclrsError>` to get a `Result<(), RclrsError>` which will reflect whether the vector contained any errors. You can see an example usage [here](https://github.com/mxgrey/ros2_rust/blob/b465f6f74eba71270e843d690e44f81ac78ca873/rclrs/src/parameter/service.rs#L446-L450).

Since the point of this PR is to introduce all the user-facing API changes that will be needed in order to support async executors (which IMO includes multi-threaded executors), I figured I should push that API change up to this PR.

mxgrey commented on rclrs/src/error.rs:
```diff
@@ -352,3 +379,80 @@ impl ToResult for rcl_ret_t {
+/// A helper trait to disregard timeouts as not an error.
+pub trait RclrsErrorFilter {
+    /// Get the first error available, or Ok(()) if there are no errors.
+    fn first_error(self) -> Result<(), RclrsError>;
```
That's both interesting and surprising to me. I knew that you could use collect to go from `Vec<Result<T, E>>` to `Result<Vec<T>, E>` but I didn't know `Vec<Result<(), E>>` to `Result<(), E>` was possible. This seems to be a special case for `()`. I can change the implementation of `fn first_error` to take advantage of this, although I think the code would be a bit less clear about what it's doing.

maspe36 replied:
We could also remove this function from the public API and let the users figure out what to do with a `Vec<Result<(), RclrsError>>`

mxgrey replied:
We're already using it [in the unit tests](https://github.com/ros2-rust/ros2_rust/pull/428/commits/b465f6f74eba71270e843d690e44f81ac78ca873#diff-3882c5aa1d3bc696162e78b8b0f429bbe0a2c82423da84eb012371774777c04bR449). I'd rather not replace it with `.into_iter().collect()` because I feel like `.into_iter().collect()` is much more abstruse than it needs to be for expressing something as simple as "give me the first error in this collection".

esteve commented:
@mxgrey 

> I think gathering all the crates up into one cargo workspace is the right solution here. How would you feel if I open up a separate PR for that?

Not sure, what would be the implications? Eventually we'd like to move the crates to separate repositories so that they can be released independently into the ROS buildfarm, though we're still far from getting there, would putting all the crates into one cargo workspace prevent that in the future?

mxgrey commented:
> Not sure, what would be the implications?

The only implication is that it you can do `$ cargo build`, `$ cargo run`, and `$ cargo test` from the root workspace, and the crates can all share one cache and one lockfile. There are no downsides as far as I can figure.

> would putting all the crates into one cargo workspace prevent that in the future?

No, there's no significant commitment to setting up a workspace. IIRC we were mostly planning on moving out `rosidl_generator_rs` which is a Python package anyway. But even if we wanted to move out `rosidl_runtime_rs` into its own repo, when it comes time to do that we would just revert the `Cargo.toml` of `rclrs` to specify a version for `rosidl_runtime_rs` (like it does right now) instead of specifying a path. But I assume the example crates will always stay in the same repo as `rclrs`, so that should continue to be useful.

esteve commented:
@mxgrey let us know when this is ready for one final review, though my only feedback was about the `create_basic_executor` function, but that's been addressed. @nwn is that change ok with you?

mxgrey commented:
@esteve I believe we've closed the loop on all feedback, so I think it's a good time to announce a final call for feedback to the working group.

nwn commented on rclrs/src/executor.rs:
```diff
@@ -1,45 +1,60 @@
-pub struct SingleThreadedExecutor {
+pub struct Executor {
+    context: Arc<ContextHandle>,
     nodes_mtx: Mutex<Vec<Weak<Node>>>,
```
With the removal of `remove_node()`, it should now be the responsibility of the `Executor` to clean up old `Node`s. So to avoid leaking the `Node` allocations, we should filter those out, possibly within `spin` or `spin_once`.

mxgrey replied:
Good point, added: [d047214](https://github.com/ros2-rust/ros2_rust/pull/428/commits/d0472147e83d91a14dd62c0a9ea6479409b9b2f0)

nwn replied:
Thanks, looks good!

nwn commented on rclrs/src/publisher.rs:
```diff
@@ -50,9 +50,9 @@ impl Drop for PublisherHandle {
 ///
-/// Sending messages does not require calling [`spin`][1] on the publisher's node.
+/// Sending messages does not require the node's executor to [spin][2].
 ///
-/// [1]: crate::spin
+/// [2]: crate::Executor::spin
```
Nit-pick, but this could just be `[1]` instead of `[2]`.

mxgrey replied:
Done: [e7f8bac](https://github.com/ros2-rust/ros2_rust/pull/428/commits/e7f8bac0b69cdefed15267473526c69fa161f125)

nwn commented:
> @mxgrey let us know when this is ready for one final review, though my only feedback was about the `create_basic_executor` function, but that's been addressed. @nwn is that change ok with you?

Yes, thanks for making the change. It just makes it less likely that we'll accidentally give special treatment to the basic executor.

esteve reviewed (approved):
@mxgrey thanks for addressing my feedback. Changes look good to me, awesome work! If there's anything down the line that needs to be tweak, we can just address it in a followup PR. Thanks! :+1:
